java-checks/src/main/java/org/sonar/java/checks/security/UserEnumerationCheck.java

```diff
@@ -0,0 +1,131 @@
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ThrowStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S5804")
+public class UserEnumerationCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Make sure allowing user enumeration is safe here.";
+  private static final String ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER = "org.springframework.security.authentication.dao.AbstractUserDetailsAuthenticationProvider";
+  private static final String USER_DETAILS_SERVICE = "org.springframework.security.core.userdetails.UserDetailsService";
+  private static final String USERNAME_NOT_FOUND_EXCEPTION = "org.springframework.security.core.userdetails.UsernameNotFoundException";
+  private static final String HIDE_USER_NOT_FOUND_EXCEPTIONS = "setHideUserNotFoundExceptions";
+  private static final String LOAD_USER_BY_USERNAME = "loadUserByUsername";
+  private static final String BOOLEAN = "boolean";
+  private static final String STRING = "java.lang.String";
+  private static final String THROWABLE = "java.lang.Throwable";
+  private static final Deque<MethodTree> STACK = new ArrayDeque<>();
+
+  private static final MethodMatchers SET_HIDE_USER_MATCHER = MethodMatchers.create()
+    .ofSubTypes(ABSTRACT_USER_DETAILS_AUTHENTICATION_PROVIDER)
+    .names(HIDE_USER_NOT_FOUND_EXCEPTIONS)
+    .addParametersMatcher(BOOLEAN)
+    .build();
+
+  private static final MethodMatchers LOAD_USER_MATCHER = MethodMatchers.create()
+    .ofSubTypes(USER_DETAILS_SERVICE)
+    .names(LOAD_USER_BY_USERNAME)
+    .addParametersMatcher(STRING)
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.METHOD_INVOCATION, Tree.Kind.THROW_STATEMENT, Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.METHOD)) {
+      STACK.push(((MethodTree) tree));
+      return;
+    }
+
+    if (tree.is(Tree.Kind.THROW_STATEMENT)) {
+      ThrowStatementTree throwStatementTree = (ThrowStatementTree) tree;
+      if (throwStatementTree.expression().symbolType().is(USERNAME_NOT_FOUND_EXCEPTION) && !isInsideLoadUserByUserName()) {
+        reportIssue(throwStatementTree.expression(), MESSAGE);
+      }
+      return;
+    }
+
+    MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
+    Arguments arguments = methodInvocationTree.arguments();
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree firstArgument = arguments.get(0);
+
+    checkHiddenUserNotFoundException(methodInvocationTree, firstArgument);
+    checkLoadUserArgUsedInExceptions(methodInvocationTree, firstArgument);
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
```

Comment:
I think I would also clear the stack when entering file, just to be sure that if the rules crash on one file (the analyzer will recover), we start over from a clear state on next file.

Reply:
is it better to do in scanFile() at the beginning or in leaveFile() at the end?

Reply:
If a rule is stateful, it's mandatory to clean the state when entering the file and it's usually done by overriding:
```
  @Override
  public void setContext(JavaFileScannerContext context) {
    mystate.clear();
    super.setContext(context);
  }
```
Because you don't know if an exception previously occurs during the last analyzed file. When an exception occurs, `leaveNode` is not called, the sate is stale, and a recovery mechanism continues with the next file.

Finally, you can also keep the cleanup in the `leaveNode` method, it frees the memory before the next file.